src/ring_theory/derivation.lean

```diff
@@ -0,0 +1,223 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Nicolò Cavalleri.
+-/
+
+import algebra.lie_algebra
+import tactic
+
+/-!
+# Derivations
+
+This file defines derivation. A derivation from the `R`-agebra `A` to the `A`-module `M` is an `R`
+linear map that satisfy the Leibniz rule.
+
+## Notation
+
+The notation `⁅D1, D2⁆` is used for the commutator of two derivations.
+
+TODO: this file is just a stub to go on with some PRs in the geometry section. It only
+implements the definition of derivations in commutative algebra. This will soon change: as soon
+as bimodules will be there in mathlib I will change this file to take into account the
+non-commutative case. Any development on the theory of derivations is discouraged until the
+definitive definition of derivation will be implemented.
+
+Also note that this file implements bundled derivations only by now. To be better integrated with
+the algebra section of mathlib it would probably be better to define a notion of `is_derivation` and
+link it to bundled derivations.
+-/
+
+open algebra ring_hom
+
+instance algebra.to_is_scalar_tower  {R : Type*} {A : Type*} [comm_semiring R] [semiring A]
+  [algebra R A] : is_scalar_tower R A A :=
+⟨λ r a b, by simp only [smul_def]; exact mul_smul _ _ _⟩
+
+section is_scalar_tower
+
+variables {R : Type*} [comm_semiring R]
+variables (A : Type*) [comm_semiring A] [algebra R A]
+variables {M : Type*} [add_comm_monoid M] [semimodule A M] [semimodule R M]
+  [is_scalar_tower R A M]
+variables {N : Type*} [add_comm_monoid N] [semimodule A N] [semimodule R N] [is_scalar_tower R A N]
+
+@[simp] lemma algebra_compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m :=
+by rw [←(one_smul A m), ←smul_assoc, algebra.smul_def, mul_one, one_smul]
+
+variable {A}
+
+@[simp] lemma algebra_compatible_smul_comm (r : R) (a : A) (m : M) : a • r • m = r • a • m :=
```

Comment:
Similarly this is similar to https://github.com/leanprover-community/mathlib/blob/c66ecd3f6d1602c08197c0971a36d1abd62deb50/src/ring_theory/algebra.lean#L1171